src/mediapipe_internal/mediapipegraphexecutor.cpp

```diff
@@ -751,14 +669,78 @@ Status receiveAndSerializePacket<mediapipe::ImageFrame>(const ::mediapipe::Packe
     HANDLE_PACKET_RECEIVAL_EXCEPTIONS();
 }
 
-Status MediapipeGraphExecutor::infer(const KFSRequest* request, KFSResponse* response, ExecutionContext executionContext, ServableMetricReporter*& reporterOut) const {
+// Two types of holders
+// One (HolderWithRequestOwnership) is required for streaming where it is OVMS who creates the request but it is not the packet type and we have to clean up
+// Second (HolderWithNoRequestOwnership) is required for unary-unary where it is gRPC who creates the request and musn't clean up
+// Specializations are for special case when the request itsef is the packet and we need to ensure there is no double free
+template <typename T>
+class HolderWithRequestOwnership : public ::mediapipe::packet_internal::Holder<T> {
+    std::shared_ptr<const KFSRequest> req;
+
+public:
+    explicit HolderWithRequestOwnership(const T* barePtr, const std::shared_ptr<const KFSRequest>& req) :
+        ::mediapipe::packet_internal::Holder<T>(barePtr),
+        req(req) {}
+};
+template <>
+class HolderWithRequestOwnership<const KFSRequest*> : public ::mediapipe::packet_internal::ForeignHolder<const KFSRequest*> {
+    const KFSRequest* hiddenPtr = nullptr;
+    std::shared_ptr<const KFSRequest> req;
+
+public:
+    explicit HolderWithRequestOwnership(const KFSRequest* barePtr, const std::shared_ptr<const KFSRequest>& req) :
+        ::mediapipe::packet_internal::ForeignHolder<const KFSRequest*>(&hiddenPtr),
+        hiddenPtr(barePtr),
+        req(req) {}
+};
+
+template <typename T>
+class HolderWithNoRequestOwnership : public ::mediapipe::packet_internal::Holder<T> {
+public:
+    explicit HolderWithNoRequestOwnership(const T* barePtr, const std::shared_ptr<const ::inference::ModelInferRequest>& req) :
+        ::mediapipe::packet_internal::Holder<T>(barePtr) {}
+};
+template <>
+class HolderWithNoRequestOwnership<const KFSRequest*> : public ::mediapipe::packet_internal::ForeignHolder<const KFSRequest*> {
+public:
+    const KFSRequest* hiddenPtr = nullptr;
+    explicit HolderWithNoRequestOwnership(const KFSRequest* barePtr, const std::shared_ptr<const KFSRequest>& req) :
+        ::mediapipe::packet_internal::ForeignHolder<const KFSRequest*>(&hiddenPtr),
+        hiddenPtr(barePtr) {}
+};
+
+template <template <typename X> typename Holder>
+static Status createPacketAndPushIntoGraph(const std::string& inputName, std::shared_ptr<const KFSRequest>& request, ::mediapipe::CalculatorGraph& graph, const ::mediapipe::Timestamp& timestamp, const stream_types_mapping_t& inputTypes) {
+    auto inputPacketType = inputTypes.at(inputName);
```

Comment:
Are we sure it will never throw?

Reply:
In case of Unary we iterate over MP input streams, in case of streaming we iterate over inputs of request but then check for existence of those inputs in input streams - that's why we have new INVALID_UNEXPECTED_INPUT status.